src/deepsparse/v2/pipeline.py

```diff
@@ -61,85 +62,66 @@ def __init__(
-            start=self.router.route[self.router.SPLIT_ROUTE],
-            end=self.router.JOIN_ROUTE,
-        )
-        inference_state_list = [
-            copy.deepcopy(inference_state) for x in range(len(batches))
-        ]
-        futures = self._scheduler_group.map(
-            batches,
-            inference_state_list,
-            func=run_with_state,
-        )
-        return self.condense_inputs([x.result() for x in futures])
+        next_step = self.router.START_ROUTE
+        operator_output = None
 
-    def _run_next_step(
-        self,
-        *args,
-        func: Callable,
-        next_step: Union[str, int],
-        input: Any = None,
-        **kwargs,
-    ):
-        """
-        Generic function to run a given func, process the output and determine the next
-        step.
-        """
-        if input:
-            operator_output = (
-                func(*args, **kwargs, **input)
-                if isinstance(input, dict)
-                else func(input, *args, **kwargs)
-            )
-        else:
-            operator_output = func(*args, **kwargs)
+        while next_step != self.router.END_ROUTE:
+            # Either a dictionary key or valid index
 
-        if isinstance(operator_output, Future):
-            operator_output = operator_output.result()
+            if next_step == self.router.SPLIT_ROUTE:
+                operator_output = self._apply_split(operator_output, inference_state)
+                next_step = self.router.route[self.router.JOIN_ROUTE]
 
-        state_update = None
-        if isinstance(operator_output, tuple):
-            state_update = operator_output[-1]
-            operator_output = operator_output[0]
+            if next_step == self.router.START_ROUTE:
+                outputs = self._run_next_step(
+                    *args,
+                    next_step=next_step,
+                    func=self._scheduler_group.submit,
+                    inference_state=inference_state,
+                    operator=self.ops[next_step],
+                    pipeline_state=pipeline_state,
+                    loop=loop,
+                    **kwargs,
+                )
+            else:
+                outputs = self._run_next_step(
+                    func=self._scheduler_group.submit,
+                    input=operator_output,
+                    next_step=next_step,
+                    inference_state=inference_state,
+                    operator=self.ops[next_step],
+                    pipeline_state=pipeline_state,
+                    loop=loop,
+                )
 
-        next_step = self.router.next(next_step, self.ops, operator_output)
-        return next_step, operator_output, state_update
+            await outputs
+            operator_output = outputs.result()
+
+            if isinstance(operator_output, tuple):
+                state_update = operator_output[-1]
+                operator_output = operator_output[0]
```

Comment:
YMMV, but there's a one-liner for this:
```suggestion
                operator_output, state_update = operator_output
```

Though I am making some assumptions about the tuple